File: Task1/Task1.js
/* Task 1
  * Choose array fn (map/filter/filterMap/some/find/findIndex) (map)
  * Prepare its callback based async counterpart
  * Prepare demo cases for the usage
*/


// callback based async counterpart
function mapWithCallback(array, callback, finalCallback) {
    const results = [];
    let processed = 0;

    array.forEach((item, index) => {
        callback(item, (error, result) => {
                if (error) {
                 console.error(error.message);
                 results[index] = null;
                } else {
                    results[index] = result;
                }

            processed++;
            if (processed === array.length) {
                finalCallback(results);
            }
        });
    });
}

// demo case for the usage
function demoFunc() {
    const numbers = [1, 2, 3, 4, 5];

    const tripleWithCallback = (num, callback) => {
        setTimeout(() => {
            if (Math.random() < 0.2) {
                callback(`Failed to process ${num}`);
            } else {
                callback(null, num * 3);
            }
        }, Math.random() * 1000);
    };

    mapWithCallback(numbers, tripleWithCallback, (results) => {
        console.log(results);
    });
}

demoFunc();











// // callback based async counterpart
// async function asyncMap(array, callback) {
//     const results = [];
//     for (const item of array) {
//         try {
//             results.push(await callback(item));
//         } catch (error) {
//             console.error(`Error processing item ${item}:`, error);
//             results.push(null);
//         }
//     }
//     return results;
// }


// // demo case for the usage
// async function demoFunc() {
//     const numbers = [1, 2, 3, 4, 5];

//     const asyncTriple = async (num) => {
//      return new Promise((resolve, reject) => {
//         setTimeout(() => {
//             if (Math.random() < 0.2) {
//                 reject(`Failed to process ${num}`);
//             } else {
//                 resolve(num * 3);
//             }
//         }, Math.random() * 1000);
//       });
//     };

//     const results = await asyncMap(numbers, asyncTriple);
//     console.log(results);
// }

// demoFunc();

File: Task1/Task1.kt
/* Task 1
  * Choose array
  * Prepare its callback based async counterpart
  * Prepare demo cases for the usage
*/

import kotlinx.coroutines.*
import kotlin.random.Random

// callback based async counterpart
suspend fun <T, R> asyncMap(list: List<T>, callback: suspend (T) -> R): List<R> {
    val results = mutableListOf<R>()
    for (item in list) {
        try {
            results.add(callback(item))
        } catch (e: Exception) {
            println("Error processing item $item: ${e.message}")
            results.add(null)
        }
    }
    return results
}

// demo case for the usage
fun main() = runBlocking {
    val numbers = listOf(1, 2, 3, 4, 5)

    val asyncTriple: suspend (Int) -> Int = { num ->
        delay(Random.nextLong(1000))
        if (Random.nextDouble() < 0.2) {
            throw Exception("Failed to process $num")
        }
        num * 3
    }

    val results = asyncMap(numbers, asyncTriple)
    println(results)
}
